blog/views.py

```diff
@@ -1,117 +1,121 @@
-
-
-@login_required
-def post_remove(request, pk):
-    """Delete post."""
-    post = get_object_or_404(Post, pk=pk)
-    post.delete()
-    return redirect('post_list')
-
-
-def add_comment(request, pk):
-    """Add comment to post."""
-    post = get_object_or_404(Post, pk=pk)
-    if request.method == 'POST':
-        form = CommentForm(request.POST)
-        if form.is_valid():
-            comment = form.save(commit=False)
-            comment.post = post
-            comment.save()
-            return redirect('post_detail', pk=post.pk)
-    else:
-        form = CommentForm()
-    return render(request, 'blog/add_comment.html', {'form': form})
-
-
-@login_required
-def comment_approve(request, pk):
-    """Approve (publish) comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    comment.approve()
-    return redirect('post_detail', pk=comment.post.pk)
-
-
-@login_required
-def comment_remove(request, pk):
-    """Delete comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    post_pk = comment.post.pk
-    comment.delete()
-    return redirect('post_detail', pk=post_pk)
+class Protected(View):
+    """Deny access to view for unauthorized users."""
+
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Call base method with login_required extension."""
+        return super().dispatch(*args, **kwargs)
+
+
+class PostListView(ListView):
+    """View for list of all published posts."""
+
+    context_object_name = 'posts'
+    template_name = 'blog/post_list.html'
+
+    def get_queryset(self):
+        """Return already published posts."""
+        return Post.objects.filter(published_date__lte=timezone.now()).order_by('published_date')
+
+
+class PostDetailView(DetailView):
+    """View with detail info about post."""
+
+    model = Post
+    template_name = 'blog/post_detail.html'
+
+
+class NewPost(CreateView, Protected):
+    """View for creating new post."""
+
+    form_class = PostForm
+    template_name = 'blog/post_edit.html'
+
+    def form_valid(self, form):
+        """Pass request.user to model."""
+        form.save(self.request.user)
+        return super(NewPost, self).form_valid(form)
+
+
+class EditPost(UpdateView, Protected):
+    """View for editing post."""
+
+    model = Post
+    fields = ['title', 'text']
+    template_name = 'blog/post_edit.html'
+
+
+class PostDraftList(ListView, Protected):
+    """View for list with unpublished posts."""
+
+    context_object_name = 'posts'
+    template_name = 'blog/post_draft_list.html'
+    queryset = Post.objects.filter(published_date__isnull=True).order_by('created_date')
+
+
+class PublishPost(UpdateView, Protected):
+    """View for publishing post."""
+
+    def get(self, request, *args, **kwargs):
+        """Publish post."""
+        post = self.get_object(Post.objects.filter(pk=kwargs['pk']))
+        post.publish()
+        return HttpResponseRedirect(reverse('post_detail', kwargs={'pk': post.pk}))
```

Comment:
If you use `UpdateView`, there's a plenty of logic encapsulated there. You should just inject a small change into some of already existing methods. It will get redirected automatically. BTW, there should be `self.object` available, which is handled so you don't have to do manual db lookup.

Reply:
self.object assigns in BaseUpdateView.get(), so I should write smthg like:
```
super.get(...)
DoSmthg (self.object.publish etc.)
return HttpResponseRedirect(...)
```

Reply:
I think you're missing the idea of pre-existing base classes. post/get/put etc. methods are low-level here. But those classes like UpdateView and similar encapsulate lots of logic and provide you with custom hook methods, which you should modify to inject some changes into existing flow.

Reply:
I think, I miss the idea, where should I implement logic in general - or just couldn't find an appropriate method to make an injection.